File: bongf/java/src/ch5_dfs_bfs/Ch5_10_FreezeDrinks.java
package ch5_dfs_bfs;

import java.util.Arrays;
import java.util.Scanner;

public class Ch5_10_FreezeDrinks {
    private static int[][] arr;
    private final static int[][] move = {{-1, 0}, {1,0}, {0, -1}, {0,1}};
    private static int n, m;

    public static void main(String[] args) {
        System.out.println(solution());
    }
    public static int solution() {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        sc.nextLine();
        arr = new int[n][m];
        for(int i=0; i<n; i++) {
            arr[i] = Arrays.stream(sc.nextLine().split(""))
                    .mapToInt(Integer::valueOf)
                    .toArray();
        }
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                // dongbin : if(dobinCheckIce()) {cnt++;} :
                if(arr[i][j] == 0) {
                    cnt++;
                    checkIce(i,j);
                }
            }
        }
        return cnt;
    }

    private static void checkIce(int i, int j) {
        arr[i][j] = 1;
        for (int[] mp : move) {
            int nx = i + mp[0];
            int ny = j + mp[1];
            if( 0<=nx && nx < n &&  0<= ny && ny < m && arr[nx][ny] == 0) {
                checkIce(nx, ny);
            }
        }
    }

    //나는 checkIce의 함수에 넣기 전에 이미 값이 유효한지 체크한다면 동빈북은 이 함수 안에서 다 체크한다.
    public static boolean dongbinCheckIce(int x, int y) {
        // 주어진 범위를 벗어나는 경우에는 즉시 종료
        if (x <= -1 || x >=n || y <= -1 || y >= m) {
            return false;
        }
        // 현재 노드를 아직 방문하지 않았다면
        if (arr[x][y] == 0) {
            // 해당 노드 방문 처리
            arr[x][y] = 1;
            // 상, 하, 좌, 우의 위치들도 모두 재귀적으로 호출
            dongbinCheckIce(x - 1, y);
            dongbinCheckIce(x, y - 1);
            dongbinCheckIce(x + 1, y);
            dongbinCheckIce(x, y + 1);
            return true;
        }
        return false;
    }
}

/*
15 14
00000111100000
11111101111110
11011101101110
11011101100000
11011111111111
11011111111100
11000000011111
01111111111111
00000000011111
01111111111000
00011111111000
00000001111000
11111111110011
11100011111111
11100011111111
결과는 8
 */

/*
4 5
00110
00011
11111
00000
결과는 3
 */

File: bongf/python/ch5_dfs_bfs/ch5_10_freeze_drinks.py
def solution() :
    n, m = map(int, input().split())
    arr = []
    for _ in range(n):
        arr.append(list(map(int, input())))
        
    move = [[-1, 0], [1,0], [0, -1], [0,1]]

    def check_ice(i, j) :
        arr[i][j] = 1
        for x,y in move :
            if( 0<=i+x<= n-1 and 0<=j+y<=m-1 and arr[i+x][j+y] == 0) :
                check_ice(i+x, j+y)

    cnt = 0 
    for i in range(n):
        for j in range(m):
            if(arr[i][j] == 0) :
                cnt += 1
                check_ice(i, j)
    return cnt

print(solution())

'''
15 14
00000111100000
11111101111110
11011101101110
11011101100000
11011111111111
11011111111100
11000000011111
01111111111111
00000000011111
01111111111000
00011111111000
00000001111000
11111111110011
11100011111111
11100011111111
'''

def dongbin() : 
    # N, M을 공백을 기준으로 구분하여 입력 받기
    n, m = map(int, input().split())

    # 2차원 리스트의 맵 정보 입력 받기
    graph = []
    for i in range(n):
        graph.append(list(map(int, input())))

    # DFS로 특정한 노드를 방문한 뒤에 연결된 모든 노드들도 방문
    def dfs(x, y):
        # 주어진 범위를 벗어나는 경우에는 즉시 종료
        if x <= -1 or x >= n or y <= -1 or y >= m:
            return False
        # 현재 노드를 아직 방문하지 않았다면
        if graph[x][y] == 0:
            # 해당 노드 방문 처리
            graph[x][y] = 1
            # 상, 하, 좌, 우의 위치들도 모두 재귀적으로 호출
            dfs(x - 1, y)
            dfs(x, y - 1)
            dfs(x + 1, y)
            dfs(x, y + 1)
            return True
        return False

    # 모든 노드(위치)에 대하여 음료수 채우기
    result = 0
    for i in range(n):
        for j in range(m):
            # 현재 위치에서 DFS 수행
            if dfs(i, j) == True:
                result += 1

    print(result) # 정답 출력
